java/core/src/main/java/com/microsoft/bond/protocol/Cloning.java

```diff
@@ -0,0 +1,54 @@
+package com.microsoft.bond.protocol;
+
+import java.io.ByteArrayInputStream;
+import java.lang.reflect.Field;
+
+class Cloning {
+    static ByteArrayInputStream clone(ByteArrayInputStream stream) {
+        // A ByteArrayInputStream can be duplicated with the three pieces of
+        // state that are passed to its long-form constructor. Unfortunately,
+        // they are hidden in protected members after construction, so we need
+        // to reflect to get them out.
+        //
+        // There's one more piece of state inside that class - mark. We don't
+        // use mark()/reset(), and cloned streams shouldn't leak out of the
+        // library, so we ignore it here.
+
+        final Field bufField;
+        final Field posField;
+        final Field countField;
+        try {
+            bufField = ByteArrayInputStream.class.getDeclaredField("buf");
```

Comment:
That's a good question. `Field.get()` doesn't promise that it can be used concurrently, and I'm having a hard-time convincing myself it can. Found a few random conversations on the internet where people state that they believe it's safe to cache your `Field` (or your `Method`, from the same package), but I can't find proof or a promise.

(see one discussion [from SO](https://stackoverflow.com/questions/5349383/is-java-lang-reflect-method-thread-safe) and another [from 2006](https://forums.manning.com/posts/list/10293.page).)

I could convince myself to do what you're asking if I dug deep enough into `java.lang.reflect`, but I'm feeling like it's not worth the time. What I have in this PR is more obviously safe and is only invoked once per `bonded`. I can imagine a large container of `bonded` or a machine-generated struct with a lot of them, but these seem like niche cases.

Your call - want me to go deep on this now, or make an issue to check it out later?